sympy/matrices/random.py

```diff
@@ -0,0 +1,2343 @@
+    Matrix([
+    [0, 1, 0],
+    [1, 0, 0],
+    [0, 0, 1]])
+
+    Parameters
+    ==========
+    dim : integer
+        dimension n of matrix
+    index : tuple(integer, integer) (optional)
+        coordinates (i,j) of transposition
+    See Also
+    ========
+    elementary
+    permutation
+
+    """
+
+    index = index or _sample(range(dim), 2)
+    return super_elementary_matrix(dim, index)
+
+
+def permutation(dim,
+                perm=None):
+    r"""permutation matrix n x n
+
+    Explamation
+    ===========
+
+    A permutation matrix is a matrix with only 0 or 1 entries
+    and each column and each row consists of only one non-zero entry.
+
+    Such a matrix can be obtained by shuffling the rows
+    of an identiy matrix .i.e ``permutation(dim, perm)`` is eqivalent to
+    ``eye(dim).permute(perm)``.
+
+    Examples
+    ========
+    >>> from sympy.core.random import seed
+    >>> from sympy.matrices.random import permutation
+    >>> seed(1)
+
+    >>> permutation(3)
+    Matrix([
+    [1, 0, 0],
+    [0, 0, 1],
+    [0, 1, 0]])
+
+    >>> permutation(3, (2,0,1))  # no more random
+    Matrix([
+    [0, 0, 1],
+    [1, 0, 0],
+    [0, 1, 0]])
+
+    Parameters
+    ==========
+    dim : integer
+        dimension n of matrix
+    perm : tuple or list of integer (optional)
+        permutation as list of n integers, e.g. ``[2,1,0,3]`` as a
+        permutation of ``[0,1,2,3]``
+
+    See Also
+    ========
+    transposition
+
+    """
+
+    perm = perm or _sample(range(dim), dim)
+    obj = zeros(dim)
+    for i, j in enumerate(perm):
+        obj[i, j] = 1
+    return obj  # aka eye(dim).permute(perm)
+
+
+def elementary(dim,
+               index=None,
+               scalar=None):
+    r"""an elementary matrix n x n for Gauss elimination
+
+    Explanantion
+    ============
+    Elementary matrices are matrices for Gauss elimination operation.
+
+    In two dimensions any matrix of the following types are elemenarty
+
+    .. math::
+
+       \mathbf{T}
+       = \left[\begin{array}{cc}0 & 1 \\ 1 & 0\end{array}\right] \\
+
+       \mathbf{M}
+       = \left[\begin{array}{cc}\lambda & 0 \\ 0 & 1\end{array}\right] \\
+
+       \mathbf{A}
+       = \left[\begin{array}{cc}1 & \mu \\ 0 & 1\end{array}\right] \\
+
+    In higher dimensions, any matrix $\mathbf{A}$
+    looking like identity matrix but with entries
+
+    .. math::
+
+        A_{ii} = a, \quad A_{ij} = b, \quad A_{ji} = -c, \quad A_{jj} = d
+
+    and $ad - bc \neq 0$.
+    So multiplication with $\mathbf{A}$ gives for
+
+    * $a=0=d$ and $b=1=-c$ a :func:`transposition`,
+      i.e. swapping rows $i$ and $j$
+    * $a=1$, $d=\lambda$ and $b=0=-c$ this matrix describes scaling the
+      row $j$ by $\lambda$
+    * $a=1=d$, $b=\mu$, $-c=0$ adding the $\mu$ multiple
+      of the row $i$ to row $j$.
+
+    Examples
+    ========
+    >>> from sympy.core.random import seed
+    >>> from sympy.matrices.random import elementary
+    >>> seed(1)
```

Comment:
I'm not sure what's the best way to handle this. Ideally the doctest wouldn't set seed like this because it would influence the behaviour of other doctests. Maybe it's better just to skip the tests that are random with `DOCTEST_SKIP`.

Reply:
Making the code invisible does not really help. the problem is that calling the seed like this alters global state. When the doctests are run this doctest will be executed and then afterwards other doctests will be executed. Ideally the test runner would be in control of the seeds for doctests so that when some doctest fails the failure is something that can be reproduced by the test runner. Also ideally we wouldn't use the same seed always when running the doctests but any doctest that runs after this one would have its seed fixed by this one.

Reply:
I though of something like storing the state of random before calling `seed()`and re-establishing the state at the end of the docstring. Since there seem no seed setting possible in `sympy.testing.runtests.doctest`.

If we could add [`sphinx.ext.doctest`](https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html#module-sphinx.ext.doctest) as sphinx framework (in `conf.py`) one could make use of [`rng.getstate()`](https://docs.sympy.org/latest/modules/core.html?highlight=random#module-sympy.core.random)  

```
    Examples
    ========

    .. testsetup::

       >>> from sympy.core.random import rng
       >>> _rng_state = rng.getstate()

    >>> from sympy import sqrt, eye, expand
    >>> from sympy.core.random import seed
    >>> from sympy.matrices.random import orthogonal
    >>> seed(1)

    >>> orthogonal(3, scalars=(-1,1))
    Matrix([
    [-1,  0, 0],
    [ 0, -1, 0],
    [ 0,  0, 1]])

    .. testcleanup::

       >>> assert not rng.getstate() == _rng_state
       >>> rng.setstate(_rng_state)
       >>> assert rng.getstate() == _rng_state

```

This seems to work fine. Wonder if this would be ok?

Otherwise I don't know better than `DOCTEST_SKIP`.

Reply:
I tried [sphinx.ext.doctest](https://github.com/sympy/sympy/pull/22103#:~:text=we%20could%20add-,sphinx.ext.doctest,-as%20sphinx%20framework) on a different [branch](https://github.com/jan-philipp-hoffmann/sympy/tree/pre_master_ext_doctest) and it seems to work well; both docs as well as doctests. The examples look more natural as invoking `seed()` could be hidden in `.. testsetup::`.

If you agree, I would merge this with this PR.

Reply:
The sphinx doctest plugin looks good but I'm not sure it's compatible with SymPy's existing doctest runner or the pytest-doctestplus plugin.

Reply:
found even better solution. We just put the hidden code into a comment like 
```
.. ..testsetup::

   >>> from sympy.core.random import rng, seed
   >>> _rng_state = rng.getstate()
   >>> seed(0)
```
so we need no additional sphinx plugin. The doc test still execute the code but it is invisible in the docs.